File: Python3.x/18-4Sum.py
# Time Limit Exceeded
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        len_nums = len(nums)
        res = []
        for a in range(len_nums - 3):
            for b in range(a + 1, len_nums - 2):
                for c in range(b + 1, len_nums - 1):
                    for d in range(c + 1, len_nums):
                        if nums[a] + nums[b] + nums[c] + nums[d] == target and [nums[a], nums[b], nums[c], nums[d]] not in res:
                            res.append([nums[a], nums[b], nums[c], nums[d]])
        return res


# space change time
# Runtime: 128 ms, faster than 77.10% of Python3 online submissions for 4Sum.
# Memory Usage: 17.4 MB, less than 8.74% of Python3 online submissions for 4Sum.
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        dic = dict()
        len_nums = len(nums)
        for i in range(len_nums):
            for j in range(i + 1, len_nums):
                sum_2 = nums[i] + nums[j]
                if sum_2 in dic:
                    dic[sum_2].append([i, j])
                else:
                    dic[sum_2] = [[i, j]]
        result = set()
        for key in dic:
            value = target - key
            if value in dic:
                list_1, list_2 = dic[key], dic[value]
                for [a, b] in list_1:
                    for [c, d] in list_2:
                        if a != c and a != d and b != c and b != d:
                            tmp = [nums[a], nums[b], nums[c], nums[d]]
                            tmp.sort()
                            result.add(tuple(tmp))
        return list(result)

File: Algorithms/C++/18-4Sum.cpp
// Runtime: 992 ms, faster than 5.02% of C++ online submissions for 4Sum.
// Memory Usage: 9.3 MB, less than 75.44% of C++ online submissions for 4Sum.
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> res;
        sort(nums.begin(), nums.end());
        for (int a = 0; a < nums.size(); ++a)
            for (int b = a + 1; b < nums.size(); ++b)
                for (int c = b + 1; c < nums.size(); ++c)
                    for (int d = c + 1; d < nums.size(); ++d)
                        if (nums[a] + nums[b] + nums[c] + nums[d] == target) {
                            vector<int> tmp = {nums[a], nums[b], nums[c], nums[d]};
                            if (find(res.begin(), res.end(), tmp) == res.end())
                                res.push_back(tmp);
                        }
        return res;
    }
};
